ioos_qc/qartod.py

```diff
@@ -611,9 +611,17 @@ def spike_test(
     flag_arr[0] = QartodFlags.UNKNOWN
     flag_arr[-1] = QartodFlags.UNKNOWN
 
-    # If the value is masked or nan set the flag to MISSING
-    flag_arr[diff.mask] = QartodFlags.MISSING
-
+    # Check if the original data was masked
+    for i in range(inp.size):
+    
+        # Check if both inp and diff are masked
+        if inp.mask[i] and diff.mask[i]:
+            flag_arr[i] = QartodFlags.MISSING
+        
+        # Check if either inp or diff is masked
+        elif inp.mask[i] or diff.mask[i]:
+            flag_arr[i] = QartodFlags.UNKNOWN
```

Comment:
Hi there, I was looped into the discussion in [issue 106](https://github.com/ioos/ioos_qc/issues/106), and @ocefpaf recommended I take a look at this PR if I had any thoughts. 

This looks good! 

Even though the resulting `flag_arr` is the same, I would think we'd want the conditionals to look like this, no?
```
       # Check if inp is masked (original data missing)
        if inp.mask[i]:
            flag_arr[i] = QartodFlags.MISSING
        
       # Check if diff is masked but not in inp (this indicates that original data is not missing, 
       # but the data point got masked in diff lines 575-580 due to trying to calculate a value 
       # using a valid value and a missing value; and because of that, we are not able to apply QARTOD
       # thus the UNKNOWN flag)
        elif (diff.mask[i] and not inp.mask[i]):
            flag_arr[i] = QartodFlags.UNKNOWN
```

@Sakshamgupta90

Reply:
Given the corrected behavior to what is expected, we will also want to change the `expected` list in `test_qartod.QartodSpikeTest.test_spike_masked()` (line ~1021) to this:

```
expected = [2, 4, 4, 4, 1, 3, 1, 2, 9, 2, 4, 4, 2, 9]
```

Previously, it was applying the MISSING FLAG 9 to the values surrounding actual missing values. Your suggested change fixes this and we should see UNKNOWN FLAG 2 surrounding missing values instead

Reply:
(similar updates need to be made to the expected outputs in `test_spike_methods()` and `test_spike_test_inputs()` I believe)

Reply:
Hi @iwensu0313 ,
I completely agree with you. The conditionals should indeed be structured as you've suggested to accurately reflect the intended behavior. 
Additionally, updating the `expected` results list is necessary to ensure the test remains valid.

Thank you for pointing out the required changes.